### PR TITLE
start compute block reward impl

### PR DESCRIPTION
**Pull request by MSghais**
- [V] issue #36 
- [V] follows contribution [guide](https://github.com/keep-starknet-strange/raito/blob/main/CONTRIBUTING.md)
- [X] code change includes tests: 
Start implementing test with error. Need to work on a test with a valid Chain state to compute the block reward.

- Start implementing compute_block_reward in the validation.cairo.
- Return a felt252, same as: https://github.com/ZeroSync/ZeroSync/blob/f1bd4736b0bac190ac48cbc8d5fcd92e2e1be7b9/src/block/block.cairo#L252.
- Adding test with a Chain state by default that's run error for the block compute reward.

### CODE REVIEW

File: src/validation.cairo [hunk 1]
```diff
@@ -70,9 +70,24 @@ fn validate_merkle_root(self: @ChainState, block: @Block) -> Result<(), ByteArra
     Result::Ok(())
 }
 
+fn compute_block_reward(self: @ChainState, block: @Block) -> Result<felt252, ByteArray> {
+    let number_halvings: u32 = *self.block_height / 210000;
+    match number_halvings {
+        0 => { return Result::Err("number_halvings equal 0"); },
+        _ => {}
+    }
+    let denominator: u32 = number_halvings*number_halvings;
```
Comment (m-kus) [inline src/validation.cairo:79]: It's actually `2**number)halvings`

You can use this snippet if you'd like https://github.com/m-kus/raito/pull/1/files#diff-6738ab9bc884db610afe82537c430edd6398401462313df53e93d7121b812964R128

File: src/validation.cairo [hunk 2]
```diff
@@ -70,9 +70,24 @@ fn validate_merkle_root(self: @ChainState, block: @Block) -> Result<(), ByteArra
     Result::Ok(())
 }
 
+fn compute_block_reward(self: @ChainState, block: @Block) -> Result<felt252, ByteArray> {
+    let number_halvings: u32 = *self.block_height / 210000;
```
[the remark this answers is not in the document]
Comment (MSghais) [reply src/validation.cairo]: My bad misunderstanding something, appreciate the links thank! 
I've made the changes

File: src/validation.cairo (hunk 1 above)
Comment (MSghais) [reply src/validation.cairo]: Appreciate thank you! I've made few changes

File: src/validation.cairo [hunk 3]
```diff
@@ -302,4 +307,41 @@ mod tests {
...
+
+
+    #[test]
+    fn test_compute_block_reward() {
+        let max_halvings: u32 = 64;
+        let reward_initial: u32 = REWARD_INITIAL.try_into().unwrap();
+        let halving_block_range = 210_000; // every 210 000 blocks
+        let mut nprevious_subsidy = REWARD_INITIAL * 2;
+        let mut loop_index: u32 = 0;
+        loop {
```
Comment (maciejka) [inline src/validation.cairo:319]: Have mercy, keep it as simple as possible. Just your function a few times in a row, compare it to the expected value.

**Review by maciejka — CHANGES_REQUESTED** [review]
Can't find compute_block_reward:
image

File: src/validation.cairo (hunk 3 above)
Comment (MSghais) [reply src/validation.cairo:333]: I've fix a merge conflict issue and clean a bit.

I redo the same test for all halvings possible like this one: 

https://github.com/bitcoin/bitcoin/blob/0f68a05c084bef3e53e3f549c403bc90b1db319c/src/test/validation_tests.cpp#L24

Comment (maciejka) [reply src/validation.cairo:333]: Please add simple test first.

File: src/validation.cairo [hunk 4]
```diff
@@ -206,9 +208,22 @@ fn validate_coinbase(block: @Block, total_fees: u256) -> Result<(), ByteArray> {
     Result::Ok(())
 }
 
+// Return BTC reward => pow to 8 to transform into Sats. Otherwise difficult to cast to u64 correctly after if needed
+fn compute_block_reward(block_height: u32) -> Result<u64, ByteArray> {
+    let number_halvings = block_height / 210_000;
+    match number_halvings {
```
Comment (maciejka) [inline src/validation.cairo:214]: This optimization should be in shr function if at all.

File: src/validation.cairo [hunk 5]
```diff
@@ -206,9 +208,22 @@ fn validate_coinbase(block: @Block, total_fees: u256) -> Result<(), ByteArray> {
...
+        0 => { return Result::Ok(REWARD_INITIAL.try_into().unwrap()); }, // return REWARD_INITAL
+        _ => {}
+    }
+    let current_reward = shr(REWARD_INITIAL, number_halvings);
```
Comment (maciejka) [inline src/validation.cairo:218]: Reward should be in sats. Here REWARD_INITIAL is 50.

File: src/validation.cairo [hunk 6]
```diff
@@ -206,9 +208,22 @@ fn validate_coinbase(block: @Block, total_fees: u256) -> Result<(), ByteArray> {
     Result::Ok(())
 }
 
+// Return BTC reward => pow to 8 to transform into Sats. Otherwise difficult to cast to u64 correctly after if needed
+fn compute_block_reward(block_height: u32) -> Result<u64, ByteArray> {
```
Comment (maciejka) [inline src/validation.cairo:212]: No need to return a Result.

**Review by maciejka — CHANGES_REQUESTED** [review]

File: src/validation.cairo (hunk 6 above)
Comment (MSghais) [reply src/validation.cairo]: I changed it to return a u64 only, thank you

File: src/validation.cairo (hunk 5 above)
Comment (MSghais) [reply src/validation.cairo]: Change to return a SATS amount, added the shl to do it

File: src/validation.cairo (hunk 3 above)
Comment (MSghais) [reply src/validation.cairo:362]: I have added few simple test too

File: src/validation.cairo [hunk 7]
```diff
@@ -302,4 +323,54 @@ mod tests {
         let result = validate_proof_of_work(@10_u256, @block);
         assert!(result.is_ok(), "Expect prev block hash lt target");
     }
+
+    // Ref implementation here:
+    // https://github.com/bitcoin/bitcoin/blob/0f68a05c084bef3e53e3f549c403bc90b1db319c/src/test/validation_tests.cpp#L24
+    #[test]
+    fn test_compute_block_reward() {
+        let max_halvings: u32 = 64;
```
Comment (maciejka) [inline src/validation.cairo:331]: We want this test to be easy to read.
Use just this test cases:
```
            (0, 50.0),                 # Genesis block
            (209999, 50.0),            # Just before first halving
            (210000, 25.0),            # First halving
            (210001, 25.0),            # Just after first halving
            (419999, 25.0),            # Just before second halving
            (420000, 12.5),            # Second halving
            (420001, 12.5),            # Just after second halving
            (629999, 12.5),            # Just before third halving
            (630000, 6.25),            # Third halving
            (630001, 6.25),            # Just after third halving
            (839999, 6.25),            # Just before fourth halving
            (840000, 3.125),           # Fourth halving
            (840001, 3.125),           # Just after fourth halving
            (1050000, 1.5625),         # Fifth halving
            (1260000, 0.78125),        # Sixth halving
            (1470000, 0.390625),       # Seventh halving
            (1680000, 0.1953125),      # Eighth halving
            (1890000, 0.09765625),     # Ninth halving
            (2100000, 0.048828125),    # Tenth halving
            (4200000, 0.000001),       # Very high block number
```

File: src/validation.cairo [hunk 8]
```diff
@@ -206,9 +208,28 @@ fn validate_coinbase(block: @Block, total_fees: u256) -> Result<(), ByteArray> {
     Result::Ok(())
 }
 
+// Return BTC reward in SATS
+fn compute_block_reward(block_height: u32) -> u64 {
```
Comment (maciejka) [inline src/validation.cairo:212]: Please implement it as one liner:
```
shr(5000000000,  block_height / 210_000).try_into().unwrap()
```

**Review by maciejka — CHANGES_REQUESTED** [review]

**Review by maciejka — APPROVED** [review]

**Pull request merged**